digest/README.md

```diff
@@ -0,0 +1,254 @@
+# digest
+
+The *digest* program first parses ethernet <u>ipv4</u> packets. Any other packets are
+rejected. After the parser recognizes an ipv4 packet, the src ip address is used as a lookup
+key for table *nh_table*. On a table hit, the programmed action *send_nh(port, srcMac, dstMac)* instance is executed.
+The action first sets the src and destination mac address, then redirects the packet to a specified port, sets the ingress_port, sets `send_digest` to true and redirects the packet. On a table miss the packet is simply dropped.
+`my_ingress_metadata_t` will hold the port on where the packet arrived and `send_digest`. `send_digest` signals to the deparser to send a digest message to user space. This digest message will hold the source mac address and the port where the packet arrived:
+
+```
+struct mac_learn_digest_t {
+    @tc_type("macaddr") bit<48> srcAddr;
+    @tc_type("dev") PortId_t ingress_port;
+    DROP_REASON drop_reason;
+};
+```
+
+## Setup Requirements
+
+Make sure that the p4node basic container has already been created at this point(as per instructions found in [p4node](https://github.com/p4tc-dev/p4tc-examples-pub.git)). To run the sample described setup here requires 4 terminals, three terminals inside the container and one on the VM side to generate traffic.
+
+### Terminal 1 (observation of tc commands on p4node).
+
+Enter the container p4node:
+
+`sudo ip netns exec p4node /bin/bash`
+
+setup path for TC binary
+
+`TC="/usr/sbin/tc"`
+
+setup the path to where the json introspection file can be found..
+
+```
+cd /home/vagrant/p4tc-examples-pub/digest/generated
+export INTROSPECTION=.
+```
+run TC monitor:
+`$TC mon p4 events`
+
+### Terminal 2 (observation of p4 digest events on p4node).
+
+First enter the container and make sure you have the introspection path setup
+
+Enter the container p4node:
+
+`sudo ip netns exec p4node /bin/bash`
+
+setup path for TC binary
+
+`TC="/usr/sbin/tc"`
+
+setup the path to where the json introspection file can be found..
+
+```
+cd /home/vagrant/p4tc-examples-pub/digest/generated
+export INTROSPECTION=.
+```
+
+run TC monitor for digest events:
+`$TC mon p4 digest`
+
+### Terminal 3 (observes incoming traffic into p4node)
+
+First enter the container and make sure you have the introspection path setup
+
+`sudo ip netns exec p4node /bin/bash`
+
+Now let's listen to traffic on port0
+
+```
+DEV=port0
+tcpdump -n -i $DEV -e
+```
+
+### Terminal 4 (to instantiate and runtime control the program)
+
+We will run commands to first load the prog and then do required runtime setup.
+
+First enter the container
+
+```
+sudo ip netns exec p4node /bin/bash
+cd /home/vagrant/p4tc-examples-pub/digest
+```
+
+```
+modprobe ext_Digest
+```
+
+Compile the parser and control blocks programs if you have not already
+
+`make`
+
```

Comment:
Shouldnt you compile first before you modprobe?

Reply:
It doesn't matter
I can modprobe before or after